ast/compile.go

```diff
@@ -3680,14 +3700,39 @@ func checkUnusedDeclaredVars(loc *Location, stack *localDeclaredVars, used VarSe
 	return errs
 }
 
-func rewriteSomeDeclStatement(g *localVarGenerator, stack *localDeclaredVars, decl *SomeDecl, errs Errors) Errors {
+func rewriteSomeDeclStatement(g *localVarGenerator, stack *localDeclaredVars, expr *Expr, errs Errors) (*Expr, Errors) {
+	e := expr.Copy()
+	decl := e.Terms.(*SomeDecl)
 	for i := range decl.Symbols {
-		v := decl.Symbols[i].Value.(Var)
-		if _, err := rewriteDeclaredVar(g, stack, v, declaredVar); err != nil {
-			errs = append(errs, NewError(CompileErr, decl.Loc(), err.Error()))
+		switch v := decl.Symbols[i].Value.(type) {
+		case Var:
+			if _, err := rewriteDeclaredVar(g, stack, v, declaredVar); err != nil {
+				return nil, append(errs, NewError(CompileErr, decl.Loc(), err.Error()))
+			}
+		case Call:
+			switch len(v) {
+			case 4: // member3
+				e.Terms = []*Term{
+					RefTerm(VarTerm(Equality.Name)),
+					v[2],
+					RefTerm(v[3], v[1]),
+				}
+			case 3: // member
+				e.Terms = []*Term{
+					RefTerm(VarTerm(Equality.Name)),
+					v[1],
+					RefTerm(v[2], NewTerm(g.Generate())),
+				}
+			}
+			for v0 := range outputVarsForExprEq(e, VarSet{}) {
+				if _, err := rewriteDeclaredVar(g, stack, v0, declaredVar); err != nil {
+					return nil, append(errs, NewError(CompileErr, decl.Loc(), err.Error()))
+				}
+			}
```

Comment:
Perhaps we should only call `rewriteDeclaredVar` for `v[1]` and `v[2]` in the `member3` case and `v1` in the `member` case. The reason being that `some x in a[i]` will rewrite `i` which I don't think is intended.

Reply:
Changed, and added tests. I think dealing with it via the safe VarSet arg is the right way forward 👇 
```
a := [
  {"x": 0, "y": "a"},
]
some { "x": i, "y": y } in a[i]
```
since the `i` in `v[1]` should also not be rewritten.